### PR TITLE
Replace fuzzer-generated PKCS #7 memory management tests

### LINKED ISSUE #6690: PKCS7 parser tests: Replace fuzzer inputs by robust test cases
The PKCS#7 unit tests include some test data that was found by fuzzing. As such it's somewhat “bizarre”, for example it uses an obscure hash (RIPEMD160), and it turns out that [they won't reach the intended function once other validation bugs are fixed](https://github.com/Mbed-TLS/mbedtls/pull/6672#issuecomment-1331394435).

The goal of this issue is to construct robust test cases for the bugs fixed by https://github.com/Mbed-TLS/mbedtls/pull/6670/commits/290f01b3f54a16045be201699becda8f500eebd5, https://github.com/Mbed-TLS/mbedtls/pull/6670/commits/e7f8c616d0b9388fd20ffd6c9730ea8188f27716 (if it's indeed a bug, if not explain why) and https://github.com/Mbed-TLS/mbedtls/pull/6670/commits/4f01121f6e598c51e42a69f3fd9a54846013117a. A robust test case should only have one thing wrong with the data, and is generally constructed by starting from a valid input and making one modification.

The test cases can be unit tests for auxiliary functions, they don't have to be for the main parser entry point.

Once we have robust test cases, we can remove the fuzzer inputs.

### PR DESCRIPTION
**Pull request by bensze01**
## Description
This PR replaces the reproducers provided by OSS-Fuzz for the memory management issues it found in PKCS#7, and adds an additional test for a code-path we touched in the course of fixing those issues.

Fixes #6690

## Gatekeeper checklist

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

### CODE REVIEW

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Did you generate the certificates and then corrupt them? If so, could you put an entry in `tests/data_files/Makefile` that does this?

For example, something like:
```sh
cat correct_cert.der | xxd -p | sed 's/4c311930/4c301930/' | xxd -rp > corrupted.der
```
Or similar command. This would make it easier to see what's going on with the testcases.

**Comment by bensze01** [comment]
@davidhorstmann-arm Will do.
FYI: I've re-checked bcc9c7481bd8080e6e323c0f68a294c38083da5a, and it doesn't affect the code coverage (other test cases exercise all of the relevant code paths), so I'm dropping it.

File: tests/suites/test_suite_pkcs7.data [hunk 1]
```diff
@@ -62,13 +62,13 @@ PKCS7 Signed Data Parse Failure Corrupt signerInfo.serial #15.2
 depends_on:MBEDTLS_SHA256_C
 pkcs7_parse:"data_files/pkcs7_signerInfo_serial_invalid_size.der":MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO
 
-pkcs7_get_signers_info_set error handling (6213931373035520)
-depends_on:MBEDTLS_RIPEMD160_C
-pkcs7_parse:"data_files/pkcs7_get_signers_info_set-missing_free-fuzz_pkcs7-6213931373035520.der":MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
+PKCS7 Signed Data Parse Fail Corrupt signerInfos[2] (6213931373035520)
```
Comment (xkqian) [inline tests/suites/test_suite_pkcs7.data:65]: Is this the order number?
I just saw a series of numbers from the beginning, #1, #2, #3... till #13,  then skipped #14 and to #15.1, #15.2, and then this number.  So what is't the special meaning of it?

**Comment by xkqian** [comment]
The branch has some conflicts, maybe you can rebase it.

**Comment by tom-cosgrove-arm** [comment]
@bensze01 Can I nag and repeat @davidhorstmann-arm's question - how were the test files generated?

And will you be resolving the conflicts, or do you want one of us to take it over?

**Comment by bensze01** [comment]
@davidhorstmann-arm @tom-cosgrove-arm I've rebased the changes and added generation steps to the Makefile

File: tests/suites/test_suite_pkcs7.data (hunk 1 above)
Comment (bensze01) [reply tests/suites/test_suite_pkcs7.data]: They used to be sequence numbers from when the PKCS7 support was originally merged - I considered removing them, but due to the truncation of test result strings, some of the tests would be indistinguishable except by their position.

File: tests/data_files/Makefile [hunk 2]
```diff
@@ -1369,6 +1369,18 @@ pkcs7_data_signed_badsigner.der: pkcs7_data_cert_signed_sha256.der
...
+# pkcs7 signature file with invalid tag in signerInfo[1].serial after long issuer name
+pkcs7_signerInfo_1_serial_invalid_tag_after_long_name.der: pkcs7_data_multiple_signed.der
+	cp $< $@
+	echo -en '\xa1' | dd of=$@ bs=1 seek=498 conv=notrunc
+all_final += pkcs7_signerInfo_1_serial_invalid_tag_after_long_name.der
+
+# pkcs7 signature file with invalid tag in signerInfo[2]
+pkcs7_signerInfo_2_invalid_tag.der: pkcs7_data_3_signed.der
+	cp $< $@
+	echo -en '\xa1' | dd of=$@ bs=1 seek=810 conv=notrunc
```
Comment (davidhorstmann-arm) [inline tests/data_files/Makefile:1381]:
```suggestion
	echo 'a1' | xxd -p -r | dd of=$@ bs=1 seek=498 conv=notrunc
all_final += pkcs7_signerInfo_1_serial_invalid_tag_after_long_name.der

# pkcs7 signature file with invalid tag in signerInfo[2]
pkcs7_signerInfo_2_invalid_tag.der: pkcs7_data_3_signed.der
	cp $< $@
	echo 'a1' | xxd -p -r | dd of=$@ bs=1 seek=810 conv=notrunc
```
I think maybe `echo` in GNU Make doesn't have the `-e` option as the original didn't work for me, even though the same command works in a regular shell.

I got a certificate that was different from the original like so (generated certificate below):
cert_diff

Using `xxd` in reverse-mode instead solves the issue for me.

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
One issue with generation, looks good otherwise.

File: tests/data_files/Makefile (hunk 2 above)
Comment (bensze01) [reply tests/data_files/Makefile:1381]: Anyway, I can swap it out for a posixly correct solution.

Comment (davidhorstmann-arm) [reply tests/data_files/Makefile:1381]: No, I'm using Ubuntu 20.04. The command `echo -en '\xa1'` works in a terminal, but when I run:
```sh
(cd tests/data_files && make pkcs7_signerInfo_1_serial_invalid_tag_after_long_name.der)
```
I get a certificate with a literal "-en \xa1" (no quotes, interestingly).

Comment (davidhorstmann-arm) [reply tests/data_files/Makefile:1381]: Ah actually I think it's a bash feature, but Makefile uses `/bin/sh` by default.

Comment (bensze01) [reply tests/data_files/Makefile:1381]: Ah, and ubuntu uses `/bin/dash` as `/bin/sh`. gotcha.

Comment (xkqian) [reply tests/data_files/Makefile:1381]: I am sorry, I tried two ways locally: 
1. With the code from Bence, and add `SHELL := /bin/bash` to Makefile under tests/data_files.
2. With code changed by David, `echo 'a1' | xxd -p -r`.  
But seems I meet one stipid issue on my local, it seems the '3f' replaces 'a1', seems a complementary code exist there...

Comment (daverodgman) [reply tests/data_files/Makefile:1381]: For a similar problem in this Makefile with a previous PR for PKCS7, I used the Make built-in `printf` instead

Comment (davidhorstmann-arm) [reply tests/data_files/Makefile:1381]: Hmm. `3f` is the ascii for `'?'`, which might imply that something in your setup is replacing any invalid unicode characters with `'?'`.

Comment (xkqian) [reply tests/data_files/Makefile:1381]: > Hmm. `3f` is the ascii for `'?'`, which might imply that something in your setup is replacing any invalid unicode characters with `'?'`.


Seems you are right, and it always output `3f `whether I change it to` 'a1', 'a2' ,'a3', ...`  But I don't remember I did the setup manually.

Comment (gilles-peskine-arm) [reply tests/data_files/Makefile:1381]: Not all implementations of `echo` support the `-e` option, and even for those who do, they don't all support the `\xNN` escape. It's a shell builtin, so it depends which shell make uses (normally `/bin/sh`) and not just which OS. On Ubuntu, `/bin/sh` is dash by default, and it supports `-e` but not `\xNN`.

You can use `printf '\241'` as a portable replacement. The downside is that you have to use octal: `\xNN` is also not portable with `printf`.

Alternatively, use `xxd`. It's not a basic Unix utility (it comes with Vim), but this makefile doesn't need to be widely portable, it only needs to run on developer machines. As a non-Vim-user, I'm perfectly fine with assuming the presence of `xxd` here. In any case `xxd` is already used.

Comment (gilles-peskine-arm) [reply tests/data_files/Makefile:1381]: @xkqian A substitution of `?` tends to happen when printing to a terminal or converting to another encoding. The commands here should only manipulate binary data. What do you see when you run `echo 'a1' | xxd -p -r | od -t x1` ?

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM

File: tests/data_files/Makefile (hunk 2 above)
Comment (xkqian) [reply tests/data_files/Makefile]: > What do you see when you run `echo 'a1' | xxd -p -r | od -t x1` ?

I can see 
"
```
0000000 a1
0000001
```
"
exactly.  Seems it's my bad.  The bytes itself has no problem, but when I open the file with `VIM`, maybe some encoding convertion happen.  Once I open the file with "`vim -b`", then everything goes well.

**Review by xkqian — APPROVED** [review]
LGTM

**Pull request merged**